dev/test/query.ts

```diff
@@ -719,6 +755,34 @@ describe('query interface', () => {
     });
   });
 
+  // Test Logical Termination on stream()
+  it('successful return without ending the stream on stream()', callback => {
+    const overrides: ApiOverride = {
+      runQuery: request => {
+        queryEquals(request);
+        return streamWithoutEnd(result('first'), result('second', true));
+      },
+    };
+
+    let counter = 0;
+    createInstance(overrides).then(firestore => {
+      const query = firestore.collection('collectionId');
+      let received = 0;
+
+      query
+        .stream()
+        .on('data', doc => {
+          expect(doc).to.be.an.instanceOf(DocumentSnapshot);
+          ++received;
+        })
+        .on('end', () => {
+          expect(received).to.equal(2);
+          expect(++counter).to.equal(1);
```

Comment:
You should test that you only receive a single 'end' callback. One think you can do is change this to:

let endCounter = 0;
query
        .stream()
        .on('data', doc => {
          expect(doc).to.be.an.instanceOf(DocumentSnapshot);
          ++receivedCounter;
        })
        .on('end', () => {
          expect(received).to.equal(2);
         ++endCounter;
          setImmediate(() => {
             expect(endCounter).to.equal(1);
             callback();
          });
        });
    });
  });
```

This way, you ensure that once all test processing is done, there is an additional check that verifies that 'end' is only called once. Your test right now does the same thing, but once you call `callback` test is considered done and it won't fail even if any expectations fail after.